File: backend/src/models/Attendance.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IAttendance extends Document {
  employee: mongoose.Types.ObjectId;
  date: Date;
  checkIn: Date;
  checkOut?: Date;
  breakTime: number;
  totalHours: number;
  status: 'present' | 'absent' | 'late' | 'half-day';
  notes?: string;
  // Approval system
  isManualEntry: boolean;
  approvalStatus: 'pending' | 'approved' | 'rejected' | 'auto-approved';
  requestedBy?: mongoose.Types.ObjectId;
  approvedBy?: mongoose.Types.ObjectId;
  approvedDate?: Date;
  rejectionReason?: string;
  // Card system integration
  cardEntryTime?: Date;
  cardExitTime?: Date;
  cardId?: string;
  entrySource: 'manual' | 'card' | 'system';
  createdAt: Date;
  updatedAt: Date;
}

const attendanceSchema = new Schema<IAttendance>({
  employee: { type: Schema.Types.ObjectId, ref: 'Employee', required: true },
  date: { type: Date, required: true },
  checkIn: { type: Date, required: true },
  checkOut: Date,
  breakTime: { type: Number, default: 0 },
  totalHours: { type: Number, default: 0 },
  status: { 
    type: String, 
    enum: ['present', 'absent', 'late', 'half-day'], 
    default: 'present' 
  },
  notes: String,
  // Approval system
  isManualEntry: { type: Boolean, default: false },
  approvalStatus: { 
    type: String, 
    enum: ['pending', 'approved', 'rejected', 'auto-approved'], 
    default: 'auto-approved' 
  },
  requestedBy: { type: Schema.Types.ObjectId, ref: 'Employee' },
  approvedBy: { type: Schema.Types.ObjectId, ref: 'Employee' },
  approvedDate: Date,
  rejectionReason: String,
  // Card system integration
  cardEntryTime: Date,
  cardExitTime: Date,
  cardId: String,
  entrySource: { 
    type: String, 
    enum: ['manual', 'card', 'system'], 
    default: 'manual' 
  }
}, { timestamps: true });

attendanceSchema.index({ employee: 1, date: 1 }, { unique: true });

export default mongoose.model<IAttendance>('Attendance', attendanceSchema);

File: rayapp/lib/models/attendance.dart
class Attendance {
  final String id;
  final String employeeId;
  final String employeeName;
  final DateTime date;
  final DateTime checkIn;
  final DateTime? checkOut;
  final double totalHours;
  final double breakTime;
  final String status;
  final String approvalStatus;
  final String entrySource;
  final String? notes;

  Attendance({
    required this.id,
    required this.employeeId,
    this.employeeName = '',
    required this.date,
    required this.checkIn,
    this.checkOut,
    required this.totalHours,
    required this.breakTime,
    required this.status,
    required this.approvalStatus,
    required this.entrySource,
    this.notes,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    final emp = json['employee'];
    final empId = emp is Map ? emp['_id'] ?? '' : emp ?? '';
    final empName = emp is Map
        ? '${emp['firstName'] ?? ''} ${emp['lastName'] ?? ''}'.trim()
        : '';
    return Attendance(
      id: json['_id'] ?? '',
      employeeId: empId,
      employeeName: empName,
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      checkIn: DateTime.tryParse(json['checkIn'] ?? '') ?? DateTime.now(),
      checkOut: json['checkOut'] != null ? DateTime.tryParse(json['checkOut']) : null,
      totalHours: (json['totalHours'] ?? 0).toDouble(),
      breakTime: (json['breakTime'] ?? 0).toDouble(),
      status: json['status'] ?? 'present',
      approvalStatus: json['approvalStatus'] ?? 'auto-approved',
      entrySource: json['entrySource'] ?? 'manual',
      notes: json['notes'],
    );
  }
}
